google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -14,64 +14,354 @@
+                raise DuplicateFirewallGroupError(
+                    'Group id already exists: %s' % group_id)
+            rule_ids = group_def.get('rule_ids')
+            if not rule_ids:
+                raise InvalidGroupDefinition(
+                    'Group "%s" does not have any rules' % group_id)
+            for rule_id in rule_ids:
+                if rule_id not in self.rules_map:
+                    raise RuleDoesntExistError(
+                        'Rule id "%s" does not exist, cannot be in group' %
+                        rule_id)
+            self.rule_groups_map[group_id] = group_def.get(
+                'rule_ids', [])
+
+    def add_org_policy(self, org_def):
+        """Creates org policy and rule mapping.
+
+        Sample org structure:
+
+                org 1234
+               /        \
+              f-1       p-c
+             /  \
+           p-a  p-b
+
+        Rules can be applied at any node above. When a policy is being audited,
+        it the rulebook will start at the lowest level (the project) and will
+        walk up the hierarchy until it reaches the first instance with rules and
+        these are the only rules that are checked.
+
+        Args:
+          org_def (dict): A dictionary of resource ids and enforced rules.
+
+        Raises:
+          RuleDoesntExistError: Raised if a rule included in the group does not
+            exist.
+          GroupDoesntExistError: Raised if a group included in an org policy
+            does not exist.
+          InvalidOrgDefinition: Raised if org policy doesn't have resources.
+        """
+        resources = org_def.get('resources', [])
+        if not resources:
+            raise InvalidOrgDefinition('Org policy does not have any resources')
+        for resource in resources:
+            resource_type = resource_mod.ResourceType.verify(
+                resource.get('type'))
+            ids = resource.get('resource_ids', [])
+            rules = resource.get('rules', {})
+            groups = rules.get('group_ids', [])
+            expanded_rules = []
+            for group in groups:
+                if group not in self.rule_groups_map:
+                    raise GroupDoesntExistError(
+                        'Group "%s" does not exist' % group)
+                expanded_group = self.rule_groups_map.get(group, [])
+                expanded_rules.extend(expanded_group)
+            for rule_id in rules.get('rule_ids', []):
+                expanded_rules.append(rule_id)
+                if rule_id not in self.rules_map:
+                    raise RuleDoesntExistError(
+                        'Rule id "%s" does not exist' % rule_id)
+            for resource_id in ids:
+                gcp_resource = resource_util.create_resource(
+                    resource_id=resource_id,
+                    resource_type=resource_type)
+                self.org_policy_rules_map[gcp_resource] = expanded_rules
```

Comment:
Once expanded_rules is a set, it would be nice to return it as a sorted list, so violations are always in the same order.

Reply:
Done.